File: tests/c/test_commutators.c
// This code is a Qiskit project.
//
// (C) Copyright IBM 2026.
//
// This code is licensed under the Apache License, Version 2.0. You may
// obtain a copy of this license in the LICENSE.txt file in the root directory
// of this source tree or at https://www.apache.org/licenses/LICENSE-2.0.
//
// Any modifications or derivative works of this code must retain this
// copyright notice, and modified files need to carry a notice indicating
// that they have been altered from the originals.

#include "common.h"
#include <qiskit_fermions.h>
#include <stdint.h>
#include <stdio.h>
#include <stdnoreturn.h>

static int test_ferm_op_commutator(void) {
    QfFermionOperator *op1 = qf_ferm_op_zero();
    QkComplex64 coeff1 = {1.0, 0.0};
    bool action1[2] = {true, false};
    uint32_t indices1[2] = {0, 0};
    qf_ferm_op_add_term(op1, 2, action1, indices1, &coeff1);
    QfFermionOperator *op2 = qf_ferm_op_zero();
    QkComplex64 coeff2 = {2.0, 0.0};
    bool action2[2] = {false, true};
    uint32_t indices2[2] = {0, 0};
    qf_ferm_op_add_term(op2, 2, action2, indices2, &coeff2);

    QfFermionOperator *comm = qf_ferm_op_commutator(op1, op2);

    QfFermionOperator *normal = qf_ferm_op_normal_ordered(comm);
    QfFermionOperator *canon = qf_ferm_op_simplify(normal, 1e-8);

    qf_ferm_op_ichop(canon, 1e-8);

    QfFermionOperator *zero = qf_ferm_op_zero();
    bool is_equal = qf_ferm_op_equal(canon, zero);

    qf_ferm_op_free(op1);
    qf_ferm_op_free(op2);
    qf_ferm_op_free(zero);

    if (!is_equal) {
        return EqualityError;
    }
    return Ok;
}

static int test_ferm_op_anti_commutator(void) {
    QfFermionOperator *op1 = qf_ferm_op_zero();
    QkComplex64 coeff1 = {1.0, 0.0};
    bool action1[2] = {true, false};
    uint32_t indices1[2] = {0, 0};
    qf_ferm_op_add_term(op1, 2, action1, indices1, &coeff1);
    QfFermionOperator *op2 = qf_ferm_op_zero();
    QkComplex64 coeff2 = {2.0, 0.0};
    bool action2[2] = {false, true};
    uint32_t indices2[2] = {0, 0};
    qf_ferm_op_add_term(op2, 2, action2, indices2, &coeff2);

    QfFermionOperator *anti_comm = qf_ferm_op_anti_commutator(op1, op2);

    QfFermionOperator *normal = qf_ferm_op_normal_ordered(anti_comm);
    QfFermionOperator *canon = qf_ferm_op_simplify(normal, 1e-8);

    qf_ferm_op_ichop(canon, 1e-8);

    QfFermionOperator *zero = qf_ferm_op_zero();
    bool is_equal = qf_ferm_op_equal(canon, zero);

    qf_ferm_op_free(op1);
    qf_ferm_op_free(op2);
    qf_ferm_op_free(zero);

    if (!is_equal) {
        return EqualityError;
    }
    return Ok;
}

static int test_ferm_op_double_commutator(void) {
    QfFermionOperator *op1 = qf_ferm_op_zero();
    QkComplex64 coeff1 = {1.0, 0.0};
    bool action1[2] = {true, false};
    uint32_t indices1[2] = {0, 0};
    qf_ferm_op_add_term(op1, 2, action1, indices1, &coeff1);
    QfFermionOperator *op2 = qf_ferm_op_zero();
    QkComplex64 coeff2 = {2.0, 0.0};
    bool action2[2] = {false, true};
    uint32_t indices2[2] = {0, 0};
    qf_ferm_op_add_term(op2, 2, action2, indices2, &coeff2);
    QfFermionOperator *op3 = qf_ferm_op_zero();
    qf_ferm_op_add_term(op3, 2, action1, indices1, &coeff1);
    QkComplex64 coeff3 = {2.0, 0.5};
    qf_ferm_op_add_term(op3, 2, action2, indices2, &coeff3);

    QfFermionOperator *double_comm = qf_ferm_op_double_commutator(op1, op2, op3, false);

    QfFermionOperator *normal = qf_ferm_op_normal_ordered(double_comm);
    QfFermionOperator *canon = qf_ferm_op_simplify(normal, 1e-8);

    qf_ferm_op_ichop(canon, 1e-8);

    QfFermionOperator *zero = qf_ferm_op_zero();
    bool is_equal = qf_ferm_op_equal(canon, zero);

    qf_ferm_op_free(op1);
    qf_ferm_op_free(op2);
    qf_ferm_op_free(zero);

    if (!is_equal) {
        return EqualityError;
    }
    return Ok;
}

int test_commutators(void) {
    int num_failed = 0;
    num_failed += RUN_TEST(test_ferm_op_commutator);
    num_failed += RUN_TEST(test_ferm_op_anti_commutator);
    num_failed += RUN_TEST(test_ferm_op_double_commutator);

    fflush(stderr);
    fprintf(stderr, "=== Number of failed subtests: %i\n", num_failed);

    return num_failed;
}

File: tests/python/operators/library/test_commutators.py
# This code is a Qiskit project.
#
# (C) Copyright IBM 2026.
#
# This code is licensed under the Apache License, Version 2.0. You may
# obtain a copy of this license in the LICENSE.txt file in the root directory
# of this source tree or at https://www.apache.org/licenses/LICENSE-2.0.
#
# Any modifications or derivative works of this code must retain this
# copyright notice, and modified files need to carry a notice indicating
# that they have been altered from the originals.

from qiskit_fermions.operators import FermionOperator
from qiskit_fermions.operators.library import (
    anti_commutator,
    commutator,
    double_commutator,
)


def test_commutator():
    op1 = FermionOperator.from_dict({((True, 0), (False, 0)): 1})
    op2 = FermionOperator.from_dict({((False, 0), (True, 0)): 2})
    comm = commutator(op1, op2)
    canon = comm.normal_ordered()
    canon.ichop()
    assert canon.equiv(FermionOperator.zero())


def test_commutator_orientation():
    """Tests that the commutator is ``AB - BA`` and not ``BA - AB``.

    The number operator obeys :math:`[n_0, a_0] = -a_0`, whose sign flips with the operand order.
    The other tests here compare against operators that either vanish or are symmetric under that
    swap, so they cannot distinguish the two.
    """
    num_op = FermionOperator.from_dict({((True, 0), (False, 0)): 1})
    ann_op = FermionOperator.from_dict({((False, 0),): 1})

    comm = commutator(num_op, ann_op).normal_ordered()

    assert comm.equiv(FermionOperator.from_dict({((False, 0),): -1}))


def test_double_commutator_nested_identity():
    """Tests the double-commutator against the nested (anti-)commutators it is defined by."""
    op_a = FermionOperator.from_dict({((True, 0), (False, 1)): 1.0, ((True, 2),): 0.5})
    op_b = FermionOperator.from_dict({((True, 1), (False, 2)): 1.0, ((False, 0),): 0.3j})
    op_c = FermionOperator.from_dict({((True, 0), (False, 2)): 0.7, ((True, 1),): 1.0 - 0.2j})

    # [[A, B], C]/2 + [A, [B, C]]/2
    expected = (
        commutator(commutator(op_a, op_b), op_c) + commutator(op_a, commutator(op_b, op_c))
    ) / 2
    actual = double_commutator(op_a, op_b, op_c, False)
    assert actual.normal_ordered().equiv(expected.normal_ordered())

    # {[A, B], C}/2 + {A, [B, C]}/2
    expected = (
        anti_commutator(commutator(op_a, op_b), op_c)
        + anti_commutator(op_a, commutator(op_b, op_c))
    ) / 2
    actual = double_commutator(op_a, op_b, op_c, True)
    assert actual.normal_ordered().equiv(expected.normal_ordered())


def test_anti_commutator():
    op1 = FermionOperator.from_dict({((True, 0), (False, 0)): 1})
    op2 = FermionOperator.from_dict({((False, 0), (True, 0)): 2})
    comm = anti_commutator(op1, op2)
    canon = comm.normal_ordered()
    canon.ichop()
    assert canon.equiv(FermionOperator.zero())


def test_double_commutator():
    op1 = FermionOperator.from_dict({((True, 0), (False, 0)): 1})
    op2 = FermionOperator.from_dict({((False, 0), (True, 0)): 2})
    op3 = FermionOperator.from_dict({((True, 0), (False, 0)): 1, ((False, 0), (True, 0)): 2 + 0.5j})
    comm = double_commutator(op1, op2, op3, False)
    canon = comm.normal_ordered()
    canon.ichop()
    assert canon.equiv(FermionOperator.zero())
